lib/workflow/stores/mongo.js

```diff
@@ -677,21 +693,91 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
 
         var query = {
             graphId: data.graphId,
-            reachable: true
-        };
-        query['dependencies.' + data.taskId] = {
-            $in: _.difference(Constants.Task.FinishedStates, [data.state])
-        };
-        var update = {
-            $set: {
-                reachable: false
-            }
+            reachable: true 
         };
+
         var options = {
             multi: true
         };
 
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        query['dependencies.'+data.taskId] = {
+            $in:_.difference(Constants.Task.FinishedStates, [data.state])
+        };
+
+       return waterline.taskdependencies.find(query).then(function(res){
+            if(_.isEmpty(res)) {
+                // try search under dependencies.anyOf.<key>
+                delete query['dependencies.' + data.taskId];
+                query['dependencies.anyOf.' + data.taskId] = {
+                    $in:_.difference(Constants.Task.FinishedStates, [data.state])
+                };
+                
+                return waterline.taskdependencies.find(query).then(function(res){
+                    if (!_.isEmpty(res)){
+                        // Mongo update, remove task under dependent.anyOf
+                        var updateRemove = {
+                            $unset:{}
+                        };
+                        updateRemove.$unset['dependencies.anyOf.'+data.taskId] = "";
+                        var queryRemove = {
+                            graphId: data.graphId,
+                            reachable: true,
+                            // List of task Ids that need to remove dependent.anyOf.<data.taskId>
+                            $or:[]
+                        };
+
+                        // Mongo update, mark the downstream task unreachable
+                        var setUnreachable = {
+                            $set: {
+                                reachable: false
+                            }
+                        };
+                        var querySetUnreachable = {
+                            graphId: data.graphId,
+                            reachable: true,
+                            // List of task Ids that need to mark as unrechable
+                            $or:[]
+                        };
+                        
+                        // res is an array of tasks that depend on data.taskId
+                        _.forEach(res, function(depItr){
+                            var depNum = Object.keys(depItr.dependencies.anyOf).length;
+                            if (depNum > 1) {
+                                var taskToRemoveDep = {};
+                                taskToRemoveDep.taskId = depItr.taskId;
+                                queryRemove.$or.push(taskToRemoveDep);
```

Comment:
Just a minor style suggestion, line 746-748 can be combined into one line and it looks much more simple:
```javascript
queryRemove.$or.push( {taskId: depItr.taskId} );
```